### PR TITLE
[Data] Make `iter_batches` an Iterable

### PR DESCRIPTION
**Pull request by amogkam**
Make `iter_batches`, `iter_torch_batches`, `iter_rows` and `iter_tf_batches` return an Iterable instead of a single use Iterator.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/iterator.py [hunk 1]
```diff
@@ -158,54 +158,105 @@ def iter_batches(
...
+                            time.perf_counter() - self.time_start
+                        )
+                    raise e
 
-        if stats:
-            stats.iter_total_s.add(time.perf_counter() - time_start)
+        block_iterator_fn = self._to_block_iterator
+        return _BatchIterator(block_iterator_fn)
 
     def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Dict[str, Any]]:
```
Comment (matthewdeng) [inline python/ray/data/iterator.py:261]: Update type hints?
```suggestion
    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterable[Dict[str, Any]]:
```

File: python/ray/data/tests/test_iterator.py [hunk 2]
```diff
@@ -158,6 +171,18 @@ def test_torch_conversion(ray_start_regular_shared):
...
+def test_torch_multi_use_iterator(ray_start_regular_shared):
+    """Tests that the iterator outputted by `iter_torch_batches` can be used
+    multiple times."""
+    ds = ray.data.range(5)
+    it = ds.iterator().iter_torch_batches()
+    it.iter_batches = MagicMock()
+
+    for batch in it:
+        assert isinstance(batch["id"], torch.Tensor)
+        assert batch["id"].tolist() == list(range(5))
```
Comment (matthewdeng) [inline python/ray/data/tests/test_iterator.py:183]: Is this the right test logic? Should it iterate through `it` multiple times instead?

Comment (amogkam) [reply python/ray/data/tests/test_iterator.py]: good point, updated

File: python/ray/data/iterator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/iterator.py:261]: this is still an Iterator

Comment (amogkam) [reply python/ray/data/iterator.py:261]: well I guess it's both

Comment (amogkam) [reply python/ray/data/iterator.py:261]: By formal definition, all iterators are also iterables.

Comment (amogkam) [reply python/ray/data/iterator.py]: updated

File: python/ray/data/tests/test_iterator.py [hunk 3]
```diff
@@ -158,6 +171,18 @@ def test_torch_conversion(ray_start_regular_shared):
...
+def test_torch_multi_use_iterator(ray_start_regular_shared):
+    """Tests that the iterator outputted by `iter_torch_batches` can be used
+    multiple times."""
+    ds = ray.data.range(5)
+    it = ds.iterator().iter_torch_batches()
+
+    for _ in range(2):
+        for batch in it:
+            assert isinstance(batch["id"], torch.Tensor)
+            assert batch["id"].tolist() == list(range(5))
```
Comment (c21) [inline python/ray/data/tests/test_iterator.py:183]: could we also test the case that iteration is stopped in the middle, without consuming all batches? This is to simulate PyTorch lightning behavior.

```
for _ in range(2):
  for i, batch in enumerate(it):
      assert ...
      if i > N:
         break
```

**Review by c21 — APPROVED** [review]
LGTM

File: doc/source/data/doc_code/tensor.py
Comment (matthewdeng) [inline doc/source/data/doc_code/tensor.py:1]: This file is no longer needed?

File: doc/source/train/doc_code/key_concepts.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def train_fn(config):
             # Local worker rank on the current machine
             "local_rank": session.get_local_rank(),
             # Data
-            "data_shard": next(dataset_shard.iter_batches(batch_format="pandas")),
+            "data_shard": next(iter(dataset_shard.iter_batches(batch_format="pandas"))),
```
Comment (matthewdeng) [inline doc/source/train/doc_code/key_concepts.py:40]: nit: This seems a bit weird to report, but it also doesn't look like this code block is used anywhere in the docs...

**Review by matthewdeng — APPROVED** [review]

File: doc/source/data/doc_code/tensor.py
Comment (amogkam) [reply doc/source/data/doc_code/tensor.py:1]: nope, it's not used anywhere in the docs

File: python/ray/data/iterator.py [hunk 5]
```diff
@@ -101,8 +102,8 @@ def iter_batches(
         _finalize_fn: Optional[Callable[[Any], Any]] = None,
         # Deprecated.
         prefetch_blocks: int = 0,
-    ) -> Iterator[DataBatch]:
-        """Return a batched iterator over the dataset.
+    ) -> Iterable[DataBatch]:
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/data/iterator.py:122]: @woshiyyya added the class here https://github.com/ray-project/ray/pull/37881/files#diff-1663692585a3befda52d637c3a1dc8183a3f69eba90f7f0a4d15fdf114a917a1R47

[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/data/iterator.py:122]: Thank you! @amogkam

**Review by woshiyyya — APPROVED** [review]

**Pull request merged**